source/state_representation/src/trajectory/JointTrajectory.cpp

```diff
@@ -0,0 +1,184 @@
+#include <algorithm>
+
+#include "state_representation/space/joint/JointState.hpp"
+#include "state_representation/trajectory/JointTrajectory.hpp"
+
+#include "state_representation/exceptions/EmptyStateException.hpp"
+#include "state_representation/exceptions/IncompatibleStatesException.hpp"
+
+namespace state_representation {
+
+JointTrajectory::JointTrajectory(const std::string& name) : TrajectoryBase<JointTrajectoryPoint>(name) {
+  this->set_type(StateType::JOINT_TRAJECTORY);
+}
+
+JointTrajectory::JointTrajectory(
+    const std::string& name, const JointState& point, const std::chrono::nanoseconds& duration
+)
+    : TrajectoryBase<JointTrajectoryPoint>(name) {
+  this->set_type(StateType::JOINT_TRAJECTORY);
+  if (point.is_empty()) {
+    throw exceptions::EmptyStateException("Point is empty");
+  }
+  this->joint_names_ = point.get_names();
+  this->add_point(point, duration);
+}
+
+JointTrajectory::JointTrajectory(
+    const std::string& name, const std::vector<JointState>& points,
+    const std::vector<std::chrono::nanoseconds>& durations
+)
+    : TrajectoryBase<JointTrajectoryPoint>(name) {
+  this->set_type(StateType::JOINT_TRAJECTORY);
+  if (points.empty()) {
+    throw exceptions::IncompatibleSizeException("No points provided");
+  } else if (std::ranges::any_of(points, [&](const auto& p) { return p.is_empty(); })) {
+    throw exceptions::EmptyStateException("Vector contains at least one point that is empty");
+  } else if (!std::ranges::all_of(points, [&](const auto& p) { return p.get_names() == points.front().get_names(); })) {
+    throw exceptions::IncompatibleStatesException("Incompatible joint names within the new points vector");
+  }
+  this->joint_names_ = points[0].get_names();
+  this->add_points(points, durations);
+}
+
+void JointTrajectory::add_point(const JointState& new_point, const std::chrono::nanoseconds& duration) {
+  if (new_point.is_empty()) {
+    throw exceptions::EmptyStateException("Point is empty");
+  } else if (this->joint_names_ != new_point.get_names()) {
```

Comment:
Does that operator work to compare all strings from a vector?

Reply:
yes, side note is that in c++20 these operators are automatically generated because of the introduction of the spaceship `<=>` (https://en.cppreference.com/w/cpp/container/vector). TL;DR we are both backwards and forwards compatible and this works